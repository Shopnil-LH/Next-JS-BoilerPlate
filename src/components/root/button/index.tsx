/**
 * Export all component defined in this folder from here
 */

import { RedirectButton } from './redirect-button';
export { RedirectButton };
