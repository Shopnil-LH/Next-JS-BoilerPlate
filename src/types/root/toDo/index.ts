export type EventType = React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement | any>

export interface IInputProps {
    type: "text" | "radio"
    placeholder: string 
    value : string
    name: string 
    onChange ?: (e:EventType ) => void 
    buttonProps: ButtonProps
}

export interface ButtonProps {
    label: string 
    onCLick ?: (e:EventType ) => void
}

export interface IListProps {
    lists: string []
}

export  interface IToDo {
    toDoTitle:string 
    inputFieldProps: IInputProps,
    lists: IListProps,
    clearHandler ?: ((e:EventType) => void)
}