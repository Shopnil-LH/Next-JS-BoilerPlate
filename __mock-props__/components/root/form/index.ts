// Internal Importstypes
import { IForm, FormElementType } from '@src/types/root';

// Internal formData
export const formData: IForm = {
  formTitle: 'Send Us Message',
  formFields: [
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'firstName',
      placeholder: 'First Name',
      isRequired: true,
      space: 0.5,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'lastName',
      placeholder: 'Last Name',
      isRequired: true,
      space: 0.5,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'email',
      placeholder: 'Enter Email Address',
      isRequired: true,
      space: 1,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'phone',
      placeholder: 'Phone number',
      isRequired: false,
      space: 1,
    },
    {
      elementType: FormElementType.TextArea,
      name: 'message',
      placeholder: 'Write Your Message Here',
      isRequired: true,
      space: 1,
      type: 'text',
    },
    {
      label: 'Terms & Condition',
      type: 'checkbox',
      elementType: FormElementType.Checkbox,
      name: 'terms',
      isRequired: true,
      space: 1,
    },
  ],
  buttonInfo: {
    buttonText: 'Send Now',
    colorScheme: 'skyblue-over-gSkyblueLight',
    size: 'btn-sm',
    'data-testid': 'submit-button',
  },
};
