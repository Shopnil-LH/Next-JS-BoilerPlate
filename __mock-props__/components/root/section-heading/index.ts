// Internal Imports
import { ISectionHeading } from '@src/types/root';

//sectionHeadingData
export const sectionHeadingData: ISectionHeading = {
  subTitle: 'Event Speakers',
  title: 'Meet Our Speakers',
  description:
    'We’re bringing together some of the best speakers in their field, and adding more as they’re confirmed.',
  isCenter: false,
  colorScheme: 'blue',
  isBottomSpacing: false,
};
