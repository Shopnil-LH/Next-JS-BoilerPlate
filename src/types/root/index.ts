import {
  colorSchemeType,
  sizeType,
  iconNameType,
  IRedirectButton,
} from './button';
import { ISectionHeading } from './section-heading';
import {
  IFormField,
  IFormValues,
  IFormFieldProp,
  IForm,
  FormElementType,
} from './form';

import {
  IToDo,
  EventType,
  IInputProps,
  ButtonProps
}  from "./toDo"

export type {
  colorSchemeType,
  sizeType,
  iconNameType,
  IRedirectButton,
  ISectionHeading,
  IFormField,
  IFormValues,
  IFormFieldProp,
  IForm,
};

export { FormElementType };

export type{
  IToDo,
  EventType,
  IInputProps,
  ButtonProps
}