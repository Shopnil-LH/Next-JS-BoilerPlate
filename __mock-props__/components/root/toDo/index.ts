import { EventType, IToDo } from "@src/types/root";

export const toDoDemoData: IToDo = {
    toDoTitle: "My Diary",
    inputFieldProps: {
        type: "text",
        placeholder: "Set Your Schedule",
        value: "mySchedule",
        name: "mySchedule",
        buttonProps: {
            label: "Add New  Schedule",
            onCLick: (e: EventType) => {alert("Clicked!!")}
        }
    },
    lists: {
        lists: ["Hello", "Bye"]
    }
}