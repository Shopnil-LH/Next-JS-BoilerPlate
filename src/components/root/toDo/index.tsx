import { EventType, IToDo } from '@src/types/root'
import React, {useState} from 'react'
import { InputData } from './inputData'
import { List } from './List'


export const ToDo = ({
    toDoTitle,
    inputFieldProps,
    lists: {
        lists
    },
    clearHandler
}: IToDo) => {
    const {
        type,
        placeholder,
        onChange,
        value,
        name,
        buttonProps: {
            label,
            onCLick
        },
        
    } = inputFieldProps
    return (
        <div className='' data-testid = "toDoHead">
            <h1 className = {`text-3xl font-bold`}>{toDoTitle || "To Do"}</h1>
            <InputData
                type = {type}
                placeholder = {placeholder || "Give me some input"}
                onChange = {onChange}
                value = {name || "random"}
                name = {name}
                buttonProps = {
                    {
                        label,
                        onCLick
                    }
                }
            />
        
            {/* show data part */}
            {/*
                If list data exist then list will be show other wise it will be hide
            */}
            {
                !!lists.length
                &&
                <div className='p-[2%] bg-red-400 inline-block rounded-[12px] mt-[2%]' >
                    <List 
                        lists= {lists}
                    />
                </div>
            }

            <br />

            {/* clear button */}
            <button 
                onClick={(e:EventType) => clearHandler ? clearHandler(e) : ():void => {}}
                className = {`px-[5%] py-[1%] bg-blue-800 text-white mt-[2%]`}
                
            >
                Clear All
            </button>
        </div>
    )
}
