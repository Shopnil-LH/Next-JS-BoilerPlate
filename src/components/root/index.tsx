import { RedirectButton } from './button';
import { SectionHeading } from './section-heading';
import { Form } from './form';
import {ToDo} from "./toDo"

export { RedirectButton, SectionHeading, Form , ToDo};
