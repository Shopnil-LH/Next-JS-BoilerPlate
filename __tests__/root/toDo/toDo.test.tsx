import { toDoDemoData } from "@root/__mock-props__/components/root/toDo"
import { ToDo } from "@src/components/root"
import { fireEvent, getByTestId, getByText, queryByTestId, render } from "@testing-library/react"
import { inputFieldDemo } from "./config"
import HomePage from "@pages/index"

/***
 * Test if the ToDo render correctly or not
 * 
 */
describe ("To Do Render Test", () => {
    it  ("render a toDo correctly", () => {
        const {getByTestId} = render (<ToDo {...toDoDemoData}/>)
        const toDo = getByTestId ("toDoHead")
        expect(toDo).toBeInTheDocument();
        expect (toDo).toMatchSnapshot();
    })
})

/***
 * 
 * Check the input field functional or not 
 */
describe("Check the toDo  input field", () => {

    it ("check input field run successfully or not", () => {
        const sampleInput = "Hello world"
        const {getByTestId} = render (<ToDo {...toDoDemoData}/>)
        const inputField = getByTestId ("toDoInput") as  HTMLInputElement
        fireEvent.change (inputField, {
            target: {
                value: sampleInput
            }
            
        })
        expect(inputField.value).toBe(sampleInput)
    })
} )


/**
 * 
 * Check the Add new Schedule button funcionble 
 */
describe ("Check the ToDo list", () => {
    let expectedOutput:string[] = [];
    let myInputItems:string [] = []
    it.each (inputFieldDemo) (`Check Input data`, (input) => {
        // expectedOutput.push (input)
        myInputItems.push (input)
        const {getAllByTestId,getByTestId} = render (
            <ToDo 
                toDoTitle= {toDoDemoData.toDoTitle}
                inputFieldProps = {toDoDemoData.inputFieldProps}
                clearHandler = {toDoDemoData.clearHandler}
                lists = {
                    {
                        lists:  myInputItems
                    }
                }
            />
        )
        const inputField = getByTestId ("toDoInput") as  HTMLInputElement
        const addButton = getByTestId ("addButton") as  HTMLInputElement
        fireEvent.change (inputField, {
            target: {
                value: input
            }
        })
        expectedOutput.push (inputField.value)
        const paragraph = getAllByTestId ("lists");
        const paraGraphElement:string[] = [];

        paragraph.forEach (p => {
            paraGraphElement.push(p.innerHTML)
        })
        // console.log(paraGraphElement)
        expect (paraGraphElement).toStrictEqual(expectedOutput)
    })
})


/**
 * 
 * Check the  new Schedule button funcionble 
 */
describe ("Check the Clear list", () => {
    it (`Check The Full TO Do`, () => {
        const myInputData = "Hello World"
        const {getByTestId, getAllByTestId, queryByTestId} = render (<HomePage/>)
        const inputField = getByTestId ("toDoInput") as  HTMLInputElement
        const addButton = getByTestId ("addButton") as  HTMLInputElement
        //change the to do input field with a static value
        fireEvent.change (inputField, {
            target: {
                value: myInputData
            }
        })
        //when user click on add button 
        fireEvent.click (addButton)
        
        const paragraph = getAllByTestId ("lists");

        //set all paragraph element in a new array
        const paraGraphElement:string[] = [];
        paragraph.forEach (p => {
            paraGraphElement.push(p.innerHTML)
        });

        //expect that paragraph data will be match with expected data
        expect (paraGraphElement).toStrictEqual ([myInputData])

        //do the clear button part
        const clearButton = getByTestId ("clearButton")
        
        //when user click on clear button
        fireEvent.click (clearButton)
        
        //expect that there don't have any paragraph with data-testid => list
        expect(queryByTestId(/list/i)).toBeNull();
    })
})