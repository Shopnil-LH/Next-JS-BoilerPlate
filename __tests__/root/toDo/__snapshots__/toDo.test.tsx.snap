// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To Do Render Test render a toDo correctly 1`] = `
<div
  class=""
  data-testid="toDoHead"
>
  <h1
    class="text-3xl font-bold"
  >
    My Diary
  </h1>
  <div
    class="space-y-3"
  >
    <label
      for="toDo_input"
    >
      Give input 
    </label>
    <input
      class="border-2 ml-5"
      data-testid="toDoInput"
      name="mySchedule"
      placeholder="Set Your Schedule"
      type="text"
    />
    <br />
    <button
      class="px-[5%] py-[1%] bg-blue-800 text-white"
      data-testid="addButton"
    >
      Add New  Schedule
    </button>
  </div>
  <div
    class="p-[2%] bg-red-400 inline-block rounded-[12px] mt-[2%]"
    data-testid="listWrapper"
  >
    <p
      data-testid="lists"
    >
      Hello
    </p>
    <p
      data-testid="lists"
    >
      Bye
    </p>
  </div>
  <br />
  <button
    class="px-[5%] py-[1%] bg-blue-800 text-white mt-[2%]"
    data-testid="clearButton"
  >
    Clear All
  </button>
</div>
`;
